build-tools/xaprepare/xaprepare/ToolRunners/GitRunner.cs

```diff
@@ -105,7 +114,14 @@ public string GetTopCommitHash (string workingDirectory = null, bool shortHash =
 				outputSink.LineCallback = (string line) => {
 					if (!String.IsNullOrEmpty (hash))
 						return;
-					hash = line?.Trim ();
+					if (Context.BuildingAZPPullRequest) {
+						// Parse merge commit message example:
+						// Merge 6d6b90779d04fb646a9bd12b4faac85285917b43 into a849e2153370453ada1bda9bc86fb98076c92498
```

Comment:
I would prefer that this use a Regex a'la `^Merge (?<hash>[0-9A-Fa-f]+) into [0-9A-Fa-f]+$` to ensure we don't accidentally match against arbitrary text, and if the match fails "fallback" to using `git rev-parse HEAD`.

Reply:
On second thought, a more simple fix may be to just immediately return the value of `Environment.GetEnvironmentVariable ("BUILD_SOURCEVERSION");` if it's not empty and we're building a pull request... 
https://docs.microsoft.com/en-us/azure/devops/pipelines/build/variables?view=azure-devops&tabs=yaml#build-variables

Reply:
`BUILD_SOURCEVERSION` is a good idea as the first choice. Regarding the regex, I don't think it's necessary - the string is well defined here so to avoid false-positive matches it would be enough to check if `messageParts[0] == "Merge"` and `messageParts[2] == "into"` which would have the same effect as a regex only would work faster.

Reply:
Unfortunately, `BUILD_SOURCEVERSION` points to a merge commit which won't be "in tree" or valid when we enter this potentially breaking state. I've reworked things a bit and updated the commit parsing based on feedback from @grendello, though we can switch over to use a Regex if that's still preferable.